diff --git a/libiberty/strncmp.c b/libiberty/strncmp.c
index 23f6df6..3c21e4e 100644
--- a/libiberty/strncmp.c
+++ b/libiberty/strncmp.c
@@ -16,19 +16,21 @@ Compares the first @var{n} bytes of two strings, returning a value as
 #include <ansidecl.h>
 #include <stddef.h>
 
-int
-strncmp(const char *s1, const char *s2, register size_t n)
-{
-  register unsigned char u1, u2;
-
-  while (n-- > 0)
-    {
-      u1 = (unsigned char) *s1++;
-      u2 = (unsigned char) *s2++;
-      if (u1 != u2)
-	return u1 - u2;
-      if (u1 == '\0')
-	return 0;
-    }
-  return 0;
-}
+extern int strncmp(const char *s1, const char *s2, register size_t n);
+
+// int
+// strncmp(const char *s1, const char *s2, register size_t n)
+// {
+//   register unsigned char u1, u2;
+
+//   while (n-- > 0)
+//     {
+//       u1 = (unsigned char) *s1++;
+//       u2 = (unsigned char) *s2++;
+//       if (u1 != u2)
+// 	return u1 - u2;
+//       if (u1 == '\0')
+// 	return 0;
+//     }
+//   return 0;
+// }
